internal/pkg/slashing/check.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/filecoin-project/go-filecoin/internal/pkg/block"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/chain"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/encoding"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/state"
+)
+
+type FaultStateView interface {
+	state.AccountStateView
+	MinerControlAddresses(ctx context.Context, maddr address.Address) (owner, worker address.Address, err error)
+}
+
+// Chain state required for checking consensus fault reports.
+type chainReader interface {
+	GetTipSet(block.TipSetKey) (block.TipSet, error)
+	FaultStateView(key block.TipSetKey) (FaultStateView, error)
+}
+
+// Checks the validity of reported consensus faults.
+type ConsensusFaultChecker struct {
+	chain chainReader
+}
+
+func NewFaultChecker(chain chainReader) *ConsensusFaultChecker {
+	return &ConsensusFaultChecker{chain: chain}
+}
+
+// Checks the validity of a consensus fault reported by serialized block headers h1, h2, and optional
+// common-ancestor witness h3.
+func (s *ConsensusFaultChecker) VerifyConsensusFault(ctx context.Context, h1, h2, extra []byte, head block.TipSetKey, earliest abi.ChainEpoch) (*runtime.ConsensusFault, error) {
+	if bytes.Equal(h1, h2) {
+		return nil, fmt.Errorf("no consensus fault: blocks identical")
+	}
+
+	var b1, b2, b3 block.Block
+	innerErr := encoding.Decode(h1, &b1)
+	if innerErr != nil {
+		return nil, errors.Wrapf(innerErr, "failed to decode h1")
+	}
+	innerErr = encoding.Decode(h2, &b2)
+	if innerErr != nil {
+		return nil, errors.Wrapf(innerErr, "failed to decode h2")
+	}
+	if len(extra) > 0 {
+		innerErr = encoding.Decode(extra, &b3)
+		if innerErr != nil {
+			return nil, errors.Wrapf(innerErr, "failed to decode extra")
+		}
+	}
+	// Block syntax is not validated. This implements the strictest check possible, and is also the simplest check
+	// possible.
+	// This means that blocks that could never have been included in the chain (e.g. with an empty parent state)
+	// are still fault-able.
+
+	if b1.Miner != b2.Miner {
+		return nil, fmt.Errorf("no consensus fault: miners differ")
+	}
+	if b1.Height > b2.Height {
+		return nil, fmt.Errorf("no consensus fault: first block is higher than second")
+	}
+
+	// Check the basic fault conditions first, defer the (expensive) signature and chain history check until last.
+	var fault *runtime.ConsensusFault
+
+	// Double-fork mining fault: two blocks at the same epoch.
+	// It is not necessary to present a common ancestor of the blocks.
+	if b1.Height == b2.Height {
+		fault = &runtime.ConsensusFault{
+			Target: b1.Miner,
+			Epoch:  b2.Height,
+			Type:   runtime.ConsensusFaultDoubleForkMining,
+		}
+	}
+	// Time-offset mining fault: two blocks with the same parent but different epochs.
+	// The height check is redundant at time of writing, but included for robustness to future changes to this method.
+	// The blocks have a common ancestor by definition (the parent).
+	if b1.Parents.Equals(b2.Parents) && b1.Height != b2.Height {
```

Comment:
@anorth I think the following is a spec regression, not a code problem but FYI:

These two checks don't seem to cover all provable forking which is a regression from previous versions of these validation rules.  Let null blocks be dashes, letter X be a tipset mined by X and A be the common ancestor of the fork in the diagrams below.  I believe these two rules cover:

Rule 1
A B
A B
and
A C D B
A T R B
etc

Rule 2
A - - - B
A - - B
etc

Neither cover
A C - B
A - - - B
because the two blocks mined by B above don't have the same parent and don't have the same epoch, however they were provable mined concurrently and we should be able to slash them.

In the past we've had forms of the slashing rule that do calculations on the parent height and block height to show an intersection in mining epoch range.  If you agree with this problem I can follow up in specs or will leave this to you as you prefer.

Reply:
To handle this properly I think the interface to the syscall would need to change, for example by including both parent headers in the extra bytes.

Reply:
Please raise an issue or discussion with the spec.

Reply:
Done: https://github.com/filecoin-project/specs/issues/891